pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go

```diff
@@ -294,7 +275,12 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 		}
 	}
 
-	if err := garbageCollection(ctx, handler); err != nil {
+	if err := handler.resourceKeeper.StateKeep(ctx); err != nil {
+		logCtx.Error(err, "Failed to run prevent-configuration-drift")
+		r.Recorder.Event(app, event.Warning(velatypes.ReasonFailedStateKeep, err))
+		return r.endWithNegativeCondition(logCtx, app, condition.ReconcileError(err), phase)
```

Comment:
What will happend In this case? I think the application is already running well here,and the health check should report error first if not? 

So , does it necessary to report an error? But a event even change the condition is necessary

Reply:
If StateKeep failed (failed to get/update target resource and is not caused by not found or conflict, such as cluster disconnect), it will trigger the next reconcile of StateKeep due to the returned error.

Reply:
I think at least a warning should be reported for users, since a failure is encountered during ensuring application managed resources and bringing state back.

Reply:
Yes， it deserves a warning. But return error just cause a new reconcile which doesn't help in this case.

Reply:
In some cases, it is worth a retry. For example, if the child cluster apiserver is busy temporarily, then the resource check might timeout which will cause a StateKeep error here. However, the child cluster apiserver might recover by itself later, then a retry will help controller do the StateKeep. Generally, it would be better to have more details for the returned error and handle it more elegantly. I agree that currently we do not necessarily need an error to be returned here. Will fix it.